### PR TITLE
Added data collator for XLNet language modeling and related calls

### LINKED ISSUE #4739: Extending run_language_modeling.py for XLNet
# 🚀 Feature request

The run_language_modeling.py script in examples/language-modeling/ currently works for BERT, RoBERTa, GPT and related models. It would be helpful if it also allowed XLNet. I believe this would involve writing functions to generate `perm_mask`, `target_mapping` and `labels` for the inputs sequences as per the paper and `https://github.com/zihangdai/xlnet/blob/master/data_utils.py`, but I'm not a 100% sure about this.  

## Motivation

I have to adapt XLNet onto a specialized domain (finance) and am yet to find a decent guide or implementation which discusses how to do this. I've decided to do this myself (with the help of this library), and would like to share my code in case others find it useful.

## Your contribution

Since I will be trying to do this anyway, I would like to submit a relevant PR when it is done. I was also hoping to receive guidance or feedback as appropriate by other contributors to ensure correctness and utility.

### PR DESCRIPTION
**Pull request by shngt**
Added `DataCollatorForXLNetLanguageModeling` in `data/data_collator.py` to return necessary inputs (applies masking and generates revelant tensors i.e. input_ids, perm_mask, target_mask and labels as per https://github.com/zihangdai/xlnet/blob/master/data_utils.py) for language modeling training with XLNetLMHeadModel. Also added related arguments, logic and calls in `examples/language-modeling/run_language_modeling.py`.

Resolves: #4739, #2008 (partially)

### CODE REVIEW

File: examples/language-modeling/run_language_modeling.py [hunk 1]
```diff
@@ -33,6 +33,7 @@
     AutoModelWithLMHead,
     AutoTokenizer,
     DataCollatorForLanguageModeling,
+    DataCollatorForXLNetLanguageModeling,
```
Comment (JetRunner) [inline examples/language-modeling/run_language_modeling.py:36]: A part of me thinks it may not be a good design. Maybe you should instead implement Permutated LM objective in the `DataCollatorForLanguageModeling` since it currently supports MLM and traditional LM (i.e., autoregressive) objective.

File: examples/language-modeling/run_language_modeling.py [hunk 2]
```diff
@@ -207,8 +209,8 @@ def main():
 
     if config.model_type in ["bert", "roberta", "distilbert", "camembert"] and not data_args.mlm:
         raise ValueError(
-            "BERT and RoBERTa-like models do not have LM heads but masked LM heads. They must be run using the --mlm "
-            "flag (masked language modeling)."
+            "BERT, RoBERTa and XLNet-like models do not have LM heads but masked LM heads. They must be run using the"
```
Comment (JetRunner) [inline examples/language-modeling/run_language_modeling.py:212]: I believe you mean "BERT, XLNet and RoBERTa-like models" since `XLM``XLMRoBERTa` are so-called RoBERTa-like models.

Comment (shngt) [reply examples/language-modeling/run_language_modeling.py:212]: Yes, that's a typo, sorry

File: examples/language-modeling/run_language_modeling.py (hunk 1 above)
Comment (shngt) [reply examples/language-modeling/run_language_modeling.py:36]: I agree; that would be a cleaner solution. Should I create a separate `--plm` flag like `--mlm` and add a new function `mask_tokens_and_perm()` like `mask_tokens()` in `DataCollatorForLanguageModeling`?

Comment (JetRunner) [reply examples/language-modeling/run_language_modeling.py:36]: @patrickvonplaten @LysandreJik What do you think?

File: src/transformers/data/data_collator.py [hunk 3]
```diff
@@ -133,3 +133,111 @@ def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]
...
+    mlm_probability: float = 0.15
+    max_gram: int = 5  # maximum individual mask span length L
+
+    def __call__(self, examples: List[torch.Tensor]) -> Dict[str, torch.Tensor]:
+        batch = self._tensorize_batch(examples)
+        if self.mlm:
+            inputs, perm_mask, target_mapping, labels = self.mask_tokens(batch)
+            return {"input_ids": inputs, "perm_mask": perm_mask, "target_mapping": target_mapping, "labels": labels}
+        else:
+            return {"input_ids": batch, "labels": batch}
```
Comment (patrickvonplaten) [inline src/transformers/data/data_collator.py:157]: do we really need the `self.mlm` parameter? I would force it to only use `mlm`. Otherwise it's the same as Causal Language Modeling and it would not make sense to use this data collator.

Comment (patrickvonplaten) [reply src/transformers/data/data_collator.py:157]: Also official it's called `Generalized Autoregressive Pretraining` so it's more CLM than MLM in my opinion. Let's try to only have one option here and give the data collator a good name

File: src/transformers/data/data_collator.py [hunk 4]
```diff
@@ -133,3 +133,111 @@ def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]
...
+    tokenizer: PreTrainedTokenizer
+    mlm: bool = True
+    mlm_probability: float = 0.15
+    max_gram: int = 5  # maximum individual mask span length L
+
+    def __call__(self, examples: List[torch.Tensor]) -> Dict[str, torch.Tensor]:
+        batch = self._tensorize_batch(examples)
+        if self.mlm:
+            inputs, perm_mask, target_mapping, labels = self.mask_tokens(batch)
+            return {"input_ids": inputs, "perm_mask": perm_mask, "target_mapping": target_mapping, "labels": labels}
```
Comment (patrickvonplaten) [inline src/transformers/data/data_collator.py:155]: We will have to pass the `mems` tensors here. Not sure yet what will be the best way to integrate this with the TF Trainer. Also pinging @TevenLeScao since he has a lot of experience training XLNet and Transfo-XL.

File: src/transformers/data/data_collator.py [hunk 5]
```diff
@@ -133,3 +133,111 @@ def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]
 
         # The rest of the time (10% of the time) we keep the masked input tokens unchanged
         return inputs, labels
+
+
+@dataclass
+class DataCollatorForXLNetLanguageModeling:
```
Comment (patrickvonplaten) [inline src/transformers/data/data_collator.py:139]: Not 100p sure what a good name here is - maybe `PermutedAutoregressiveLanuageModeling` or we stick to `GeneralizedAutoregressiveLanguageModeling` or `PermutationSampledAutoregressiveLM` ...? What do you think is a good name? Would prefer to not make it model specific even though only XLNet does this kind of training at the moment. @thomwolf @TevenLeScao - do you have a good idea for a name?

File: examples/language-modeling/run_language_modeling.py [hunk 6]
```diff
@@ -221,9 +223,17 @@ def main():
 
     train_dataset = get_dataset(data_args, tokenizer=tokenizer) if training_args.do_train else None
     eval_dataset = get_dataset(data_args, tokenizer=tokenizer, evaluate=True) if training_args.do_eval else None
-    data_collator = DataCollatorForLanguageModeling(
-        tokenizer=tokenizer, mlm=data_args.mlm, mlm_probability=data_args.mlm_probability
-    )
+    if config.model_type == "xlnet":
+        data_collator = DataCollatorForXLNetLanguageModeling(
```
Comment (patrickvonplaten) [inline examples/language-modeling/run_language_modeling.py:227]: I'm ok with this -> what do you think @julien-c @LysandreJik ?

File: examples/language-modeling/run_language_modeling.py (hunk 1 above)
Comment (patrickvonplaten) [reply examples/language-modeling/run_language_modeling.py:36]: I actually prefer the way it is done now (Creating a completely new data collator for XLNet). Also pinging @julien-c to hear his opinion on this

File: src/transformers/data/data_collator.py [hunk 7]
```diff
@@ -133,3 +133,111 @@ def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]
...
+        else:
+            if self.tokenizer._pad_token is None:
+                raise ValueError(
+                    "You are attempting to pad samples but the tokenizer you are using"
+                    f" ({self.tokenizer.__class__.__name__}) does not have one."
+                )
+            return pad_sequence(examples, batch_first=True, padding_value=self.tokenizer.pad_token_id)
+
+    def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
+        """
```
Comment (patrickvonplaten) [inline src/transformers/data/data_collator.py:173]: That's really clean! Good job!

**Comment by patrickvonplaten** [comment]
Thanks a lot for the PR @shngt - this looks really cool :-) 

It's won't actually be that easy to make this work with the trainer since we will have to pass `mems` to the `model(...)` and retrieve it from the outputs. This won't be solved in a nice clean way IMO. 

We will have to adapt this function here: https://github.com/huggingface/transformers/blob/331d8d2936e7a140225cf60301ba6469930fd216/src/transformers/trainer.py#L572

Because we don't have `namedtuples` (another case why we should have `namedtuples` :D @thomwolf @LysandreJik) 
I would suggest to add the following lines here: https://github.com/huggingface/transformers/blob/331d8d2936e7a140225cf60301ba6469930fd216/src/transformers/trainer.py#L580 for the moment (not very pretty to this :-/):

```python 
        outputs = model(**inputs)
        loss = outputs[0]  # model outputs are always tuple in transformers (see doc)
        
        ...
        if model.config_class.__name__ in ["XLNetConfig", "TransfoXLConfig"]:
            mems = outputs[1]
        else:
            mems = None
       return loss, mems
```

I don't see a better solution at the moment. What do you think @LysandreJik @julien-c ?

Would it be ok for you to add a couple of "not-so-pretty" if-statements to the trainer? 

@shngt - Let's first decide on how to adapt the `Trainer` before continuing.

**Comment by patrickvonplaten** [comment]
Wait for #5399 to be merged :-)

File: src/transformers/data/data_collator.py (hunk 5 above)
Comment (shngt) [reply src/transformers/data/data_collator.py:139]: The authors have named the objective `PermutationLanguageModeling` (section 2.2 in the paper) - shall we stick with that?

File: src/transformers/data/data_collator.py (hunk 3 above)
Comment (shngt) [reply src/transformers/data/data_collator.py:157]: No, I don't think this is necessary - I was just mimicking how the primary DataCollator works. I'll remove it my next commit and add lines to catch if `self.mlm` is not provided.

File: src/transformers/data/data_collator.py (hunk 7 above)
Comment (shngt) [reply src/transformers/data/data_collator.py:173]: Thanks so much! :)

**Comment by shngt** [comment]
Sorry for the mess, I didn't know rebasing would do this. Repeating the commit message for clarity:
Changed the name of `DataCollatorForXLNetLanguageModeling` to the more general `DataCollatorForPermutationLanguageModelling`.
Removed the `--mlm` flag requirement for the new collator and defined a separate `--plm_probability` flag for its use.
CTRL uses a CLM loss just like GPT and GPT-2, so should work out of the box with this script (provided `past` is taken care of
similar to `mems` for XLNet). Added a few words in the comments to reflect this.
Changed calls and imports appropriately.

**Comment by patrickvonplaten** [comment]
Yeah rebasing can be dangerous, happens to us all the time :D. 
#5399 is merged so it should be possible to add the XLNet data collator.

I would here to copy the only file that is changed from the branch `src/transformers/data/data_collator.py` (I tihnk) and then open a new PR where you paste in this file. Then this PR should be clean. We can close this PR then since it seems to be borked.
